stablehlo/dialect/VhloOps.td

```diff
@@ -0,0 +1,957 @@
+  let results = (outs VHLO_Tensor);
+}
+
+def VHLO_BroadcastOpV1 : VHLO_Op<"broadcast"> {
+  let arguments = (ins
+    VHLO_Tensor:$operand,
+    I64ElementsAttr:$broadcast_sizes
+  );
+  let results = (outs VHLO_Tensor);
+}
+
+def VHLO_BroadcastInDimOpV1 : VHLO_Op<"broadcast_in_dim"> {
+  let arguments = (ins
+    VHLO_Tensor:$operand,
+    BroadcastDimAttr:$broadcast_dimensions
+  );
+  let results = (outs VHLO_StaticShapeTensor);
+}
+
+def VHLO_DynamicBroadcastInDimOpV1 : VHLO_Op<"dynamic_broadcast_in_dim"> {
+  let arguments = (ins
+    VHLO_Tensor:$operand,
+    VHLO_DimensionTensor:$output_dimensions,
+    BroadcastDimAttr:$broadcast_dimensions,
+    OptionalAttr<BroadcastDimAttr>:$known_expanding_dimensions,
+    OptionalAttr<BroadcastDimAttr>:$known_nonexpanding_dimensions
+  );
+  let results = (outs VHLO_Tensor);
+}
+
+def VHLO_CholeskyOpV1 : VHLO_Op<"cholesky"> {
+  let arguments = (ins
+    VHLO_FpOrComplexTensor:$a,
+    DefaultValuedOptionalAttr<BoolAttr, "false">:$lower
+  );
+  let results = (outs VHLO_FpOrComplexTensor:$result);
+}
+
+def VHLO_ClampOpV1 : VHLO_Op<"clamp"> {
+  let arguments = (ins
+    VHLO_Tensor:$min,
+    VHLO_Tensor:$operand,
+    VHLO_Tensor:$max
+  );
+  let results = (outs VHLO_Tensor:$result);
+}
+
+def VHLO_ConcatenateOpV1 : VHLO_Op<"concatenate"> {
+  let arguments = (ins
+    Variadic<VHLO_Tensor>:$inputs,
+    I64Attr:$dimension
+  );
+  let results = (outs VHLO_Tensor);
+}
+
+def VHLO_CollectivePermuteOpV1 : VHLO_Op<"collective_permute", "0.3.0", "0.3.0"> {
+  let arguments = (ins
+    VHLO_Tensor:$operand,
+    I64ElementsAttr:$source_target_pairs
+  );
+  let results = (outs VHLO_Tensor);
+}
+
+def VHLO_CollectivePermuteOpV2 : VHLO_Op<"collective_permute_v2", "0.4.0"> {
+  let arguments = (ins
+    VHLO_Tensor:$operand,
+    I64ElementsAttr:$source_target_pairs,
+    OptionalAttr<VHLO_ChannelHandle>:$channel_handle
+  );
+  let results = (outs VHLO_Tensor);
+}
+
+def VHLO_ConvolutionOpV1 : VHLO_Op<"convolution"> {
+  let arguments = !con(
+    (ins
+       VHLO_Tensor:$lhs,
+       VHLO_Tensor:$rhs),
+    VHLO_ConvolutionAttributes.attributes);
+  let results = (outs VHLO_Tensor);
+}
+
+def VHLO_CrossReplicaSumOpV1 : VHLO_Op<"cross-replica-sum"> {
+  let arguments = (ins
+    VHLO_Tensor:$operand,
+    I64ElementsAttr:$replica_groups
+  );
+  let results = (outs VHLO_Tensor);
+}
+
+def VHLO_CustomCallOpV1 : VHLO_Op<"custom_call", "0.3.0", "0.3.0"> {
+  let arguments = (ins
+    Variadic<VHLO_TensorOrTokenOrTuple>:$inputs,
+    StrAttr:$call_target_name,
+    DefaultValuedOptionalAttr<BoolAttr, "false">:$has_side_effect,
+    DefaultValuedStrAttr<StrAttr, "">:$backend_config,
+    DefaultValuedOptionalAttr<
+        VHLO_CustomCallApiVersionAttr,
+        "::mlir::vhlo::CustomCallApiVersion::API_VERSION_ORIGINAL">:
+        $api_version,
+    DefaultValuedOptionalAttr<VHLO_FlatSymbolRefArrayAttr, "{}">:$called_computations,
+    OptionalAttr<VHLO_ArrayOfLayoutAttr>:$operand_layouts
```

Comment:
This should also include result_layouts, and the patterns/tests will need to be updated.

Reply:
Fixed. OutputOperandAlias is a strange one because it is both Optional and DefaultValued.

I realized the policy for adding default valued attrs should be "if the value is the default value, we can downgrade," unless the default is a new behavior, which isn't the case here.

Meaning both of the following should be valid downgrades:

```
func.func @main(%arg0 : tensor<f32>) -> tensor<f32> {
  %0 = stablehlo.custom_call @foo(%arg0) {
    operand_layouts = [dense<> : tensor<0xindex>],
    result_layouts = [dense<> : tensor<0xindex>]
  } : (tensor<f32>) -> tensor<f32>
  %1 = stablehlo.custom_call @foo(%arg0) {
    operand_layouts = [dense<> : tensor<0xindex>],
    output_operand_aliases = [],
    result_layouts = [dense<> : tensor<0xindex>]
  } : (tensor<f32>) -> tensor<f32>
  func.return %0 : tensor<f32>
}
```

(one without output_operand_alias, one with the default value of output_operand_alias)